File: src/clients/desktop/ReelRoulette.DesktopApp/LibraryProjectionDisplayFilter.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace ReelRoulette;

/// <summary>
/// Applies <see cref="FilterState"/> to the in-memory server library projection for desktop UI only.
/// Playback eligibility and random picks are API-authoritative; this mirrors filter rules so the library panel
/// matches user expectations without referencing server-side <c>FilterSetBuilder</c> types.
/// </summary>
public static class LibraryProjectionDisplayFilter
{
    public static bool PassesFilterState(LibraryItem item, LibraryIndex libraryIndex, FilterState filterState)
    {
        ArgumentNullException.ThrowIfNull(item);
        ArgumentNullException.ThrowIfNull(libraryIndex);
        ArgumentNullException.ThrowIfNull(filterState);

        var enabledSourceIds = libraryIndex.Sources
            .Where(s => s.IsEnabled)
            .Select(s => s.Id)
            .ToHashSet(StringComparer.OrdinalIgnoreCase);
        if (!enabledSourceIds.Contains(item.SourceId))
        {
            return false;
        }

        if (filterState.IncludedSourceIds is { Count: > 0 })
        {
            var included = filterState.IncludedSourceIds
                .Where(id => !string.IsNullOrWhiteSpace(id))
                .ToHashSet(StringComparer.OrdinalIgnoreCase);
            if (!included.Contains(item.SourceId))
            {
                return false;
            }
        }

        if (filterState.ExcludeBlacklisted && item.IsBlacklisted)
        {
            return false;
        }

        if (filterState.FavoritesOnly && !item.IsFavorite)
        {
            return false;
        }

        if (filterState.OnlyNeverPlayed && item.PlayCount != 0)
        {
            return false;
        }

        if (filterState.AudioFilter == AudioFilterMode.WithAudioOnly)
        {
            if (item.MediaType != MediaType.Photo && item.HasAudio != true)
            {
                return false;
            }
        }
        else if (filterState.AudioFilter == AudioFilterMode.WithoutAudioOnly)
        {
            if (item.MediaType != MediaType.Photo && item.HasAudio != false)
            {
                return false;
            }
        }

        if (filterState.MinDuration.HasValue)
        {
            if (item.MediaType != MediaType.Photo &&
                (!item.Duration.HasValue || item.Duration.Value < filterState.MinDuration.Value))
            {
                return false;
            }
        }

        if (filterState.MaxDuration.HasValue)
        {
            if (item.MediaType != MediaType.Photo &&
                (!item.Duration.HasValue || item.Duration.Value > filterState.MaxDuration.Value))
            {
                return false;
            }
        }

        if (filterState.OnlyKnownDuration && item.MediaType != MediaType.Photo && !item.Duration.HasValue)
        {
            return false;
        }

        if (filterState.OnlyKnownLoudness && item.MediaType != MediaType.Photo && !item.IntegratedLoudness.HasValue)
        {
            return false;
        }

        if (filterState.SelectedTags is { Count: > 0 } && !PassesSelectedTags(item, libraryIndex, filterState))
        {
            return false;
        }

        if (filterState.ExcludedTags is { Count: > 0 })
        {
            if (item.Tags != null &&
                filterState.ExcludedTags.Any(ex =>
                    item.Tags.Any(t => string.Equals(t, ex, StringComparison.OrdinalIgnoreCase))))
            {
                return false;
            }
        }

        if (filterState.MediaTypeFilter == MediaTypeFilter.VideosOnly && item.MediaType != MediaType.Video)
        {
            return false;
        }

        if (filterState.MediaTypeFilter == MediaTypeFilter.PhotosOnly && item.MediaType != MediaType.Photo)
        {
            return false;
        }

        return true;
    }

    private static bool PassesSelectedTags(LibraryItem item, LibraryIndex libraryIndex, FilterState filterState)
    {
        var selected = filterState.SelectedTags;
        if (selected == null || selected.Count == 0)
        {
            return true;
        }

        var hasCategories = libraryIndex.Categories is { Count: > 0 };
        if (!hasCategories)
        {
            return PassesLegacyTagFilter(item, filterState);
        }

        var tags = libraryIndex.Tags ?? new List<Tag>();
        var tagsByCategory = new Dictionary<string, List<string>>(StringComparer.OrdinalIgnoreCase);
        foreach (var selectedTag in selected)
        {
            var tag = tags.FirstOrDefault(t => string.Equals(t.Name, selectedTag, StringComparison.OrdinalIgnoreCase));
            if (tag != null)
            {
                if (!tagsByCategory.ContainsKey(tag.CategoryId))
                {
                    tagsByCategory[tag.CategoryId] = new List<string>();
                }

                tagsByCategory[tag.CategoryId].Add(selectedTag);
            }
            else
            {
                const string uncategorizedId = "";
                if (!tagsByCategory.ContainsKey(uncategorizedId))
                {
                    tagsByCategory[uncategorizedId] = new List<string>();
                }

                tagsByCategory[uncategorizedId].Add(selectedTag);
            }
        }

        if (tagsByCategory.Count == 0)
        {
            return true;
        }

        var categoryOutcomes = new List<bool>();
        foreach (var kvp in tagsByCategory)
        {
            var categoryId = kvp.Key;
            var categoryTags = kvp.Value;
            var localMode = filterState.CategoryLocalMatchModes?.TryGetValue(categoryId, out var m) == true
                ? m
                : TagMatchMode.And;

            bool catMatch;
            if (localMode == TagMatchMode.And)
            {
                catMatch = item.Tags != null &&
                           categoryTags.All(tag => item.Tags.Any(it =>
                               string.Equals(it, tag, StringComparison.OrdinalIgnoreCase)));
            }
            else
            {
                catMatch = item.Tags != null &&
                           categoryTags.Any(tag => item.Tags.Any(it =>
                               string.Equals(it, tag, StringComparison.OrdinalIgnoreCase)));
            }

            categoryOutcomes.Add(catMatch);
        }

        if (categoryOutcomes.Count == 0)
        {
            return false;
        }

        var useAndBetweenCategories = filterState.GlobalMatchMode ?? true;
        return useAndBetweenCategories
            ? categoryOutcomes.All(x => x)
            : categoryOutcomes.Any(x => x);
    }

    private static bool PassesLegacyTagFilter(LibraryItem item, FilterState filterState)
    {
        if (item.Tags == null)
        {
            return false;
        }

        if (filterState.TagMatchMode == TagMatchMode.And)
        {
            return filterState.SelectedTags.All(tag =>
                item.Tags!.Any(it => string.Equals(it, tag, StringComparison.OrdinalIgnoreCase)));
        }

        return filterState.SelectedTags.Any(tag =>
            item.Tags!.Any(it => string.Equals(it, tag, StringComparison.OrdinalIgnoreCase)));
    }
}

File: src/clients/web/ReelRoulette.WebUI/src/library/libraryProjectionDisplayFilter.ts
import {
  AUDIO_FILTER,
  MEDIA_TYPE_FILTER,
  TAG_MATCH_MODE,
  type FilterState,
  type TagMatchMode
} from "../filter/filterStateModel";
import type { LibraryProjectionCatalog, LibraryProjectionItem } from "./libraryProjectionModel";

function tagEquals(a: string, b: string): boolean {
  return a.localeCompare(b, undefined, { sensitivity: "accent" }) === 0;
}

function enabledSourceIds(catalog: LibraryProjectionCatalog): Set<string> {
  const ids = new Set<string>();
  for (const source of catalog.sources) {
    if (source.isEnabled) {
      ids.add(source.id);
    }
  }
  return ids;
}

function passesLegacyTagFilter(item: LibraryProjectionItem, filterState: FilterState): boolean {
  if (!item.tags.length) {
    return false;
  }
  if (filterState.tagMatchMode === TAG_MATCH_MODE.And) {
    return filterState.selectedTags.every((tag) => item.tags.some((it) => tagEquals(it, tag)));
  }
  return filterState.selectedTags.some((tag) => item.tags.some((it) => tagEquals(it, tag)));
}

function passesSelectedTags(
  item: LibraryProjectionItem,
  catalog: LibraryProjectionCatalog,
  filterState: FilterState
): boolean {
  const selected = filterState.selectedTags;
  if (!selected.length) {
    return true;
  }

  if (!catalog.categories.length) {
    return passesLegacyTagFilter(item, filterState);
  }

  const tagsByCategory = new Map<string, string[]>();
  for (const selectedTag of selected) {
    const tag = catalog.tags.find((t) => tagEquals(t.name, selectedTag));
    const categoryId = tag ? tag.categoryId : "";
    const bucket = tagsByCategory.get(categoryId) ?? [];
    bucket.push(selectedTag);
    tagsByCategory.set(categoryId, bucket);
  }

  if (tagsByCategory.size === 0) {
    return true;
  }

  const categoryOutcomes: boolean[] = [];
  for (const [categoryId, categoryTags] of tagsByCategory) {
    const localMode: TagMatchMode =
      filterState.categoryLocalMatchModes?.[categoryId] ?? TAG_MATCH_MODE.And;

    let catMatch: boolean;
    if (localMode === TAG_MATCH_MODE.And) {
      catMatch =
        item.tags.length > 0 &&
        categoryTags.every((tag) => item.tags.some((it) => tagEquals(it, tag)));
    } else {
      catMatch =
        item.tags.length > 0 &&
        categoryTags.some((tag) => item.tags.some((it) => tagEquals(it, tag)));
    }
    categoryOutcomes.push(catMatch);
  }

  if (!categoryOutcomes.length) {
    return false;
  }

  const useAndBetweenCategories = filterState.globalMatchMode ?? true;
  return useAndBetweenCategories ? categoryOutcomes.every(Boolean) : categoryOutcomes.some(Boolean);
}

export function passesFilterState(
  item: LibraryProjectionItem,
  catalog: LibraryProjectionCatalog,
  filterState: FilterState
): boolean {
  const enabled = enabledSourceIds(catalog);
  if (!enabled.has(item.sourceId)) {
    return false;
  }

  if (filterState.includedSourceIds.length > 0) {
    const included = new Set(filterState.includedSourceIds.filter(Boolean).map((id) => id.toLowerCase()));
    if (!included.has(item.sourceId.toLowerCase())) {
      return false;
    }
  }

  if (filterState.excludeBlacklisted && item.isBlacklisted) {
    return false;
  }

  if (filterState.favoritesOnly && !item.isFavorite) {
    return false;
  }

  if (filterState.onlyNeverPlayed && item.playCount !== 0) {
    return false;
  }

  if (filterState.audioFilter === AUDIO_FILTER.WithAudioOnly) {
    if (item.mediaType !== "photo" && item.hasAudio !== true) {
      return false;
    }
  } else if (filterState.audioFilter === AUDIO_FILTER.WithoutAudioOnly) {
    if (item.mediaType !== "photo" && item.hasAudio !== false) {
      return false;
    }
  }

  if (filterState.minDurationSeconds != null && Number.isFinite(filterState.minDurationSeconds)) {
    if (
      item.mediaType !== "photo" &&
      (item.durationSeconds == null || item.durationSeconds < filterState.minDurationSeconds)
    ) {
      return false;
    }
  }

  if (filterState.maxDurationSeconds != null && Number.isFinite(filterState.maxDurationSeconds)) {
    if (
      item.mediaType !== "photo" &&
      (item.durationSeconds == null || item.durationSeconds > filterState.maxDurationSeconds)
    ) {
      return false;
    }
  }

  if (filterState.onlyKnownDuration && item.mediaType !== "photo" && item.durationSeconds == null) {
    return false;
  }

  if (filterState.onlyKnownLoudness && item.mediaType !== "photo" && item.integratedLoudness == null) {
    return false;
  }

  if (filterState.selectedTags.length > 0 && !passesSelectedTags(item, catalog, filterState)) {
    return false;
  }

  if (filterState.excludedTags.length > 0) {
    if (filterState.excludedTags.some((ex) => item.tags.some((t) => tagEquals(t, ex)))) {
      return false;
    }
  }

  if (filterState.mediaTypeFilter === MEDIA_TYPE_FILTER.VideosOnly && item.mediaType !== "video") {
    return false;
  }

  if (filterState.mediaTypeFilter === MEDIA_TYPE_FILTER.PhotosOnly && item.mediaType !== "photo") {
    return false;
  }

  return true;
}

export function filterItemsByFilterState(
  items: readonly LibraryProjectionItem[],
  catalog: LibraryProjectionCatalog,
  filterState: FilterState
): LibraryProjectionItem[] {
  return items.filter((item) => passesFilterState(item, catalog, filterState));
}
